### PR TITLE
[SPARK-42440][CONNECT] Initial set of Dataframe APIs for Scala Client

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
Add a lot of the existing Dataframe APIs to the Spark Connect Scala Client.

This PR does not contain:
- Typed APIs
- Aggregation
- Streaming (not supported by connect just yet)
- NA/Stats functions
- TempView registration.

### Why are the changes needed?
We want the Scala Client Dataset to reach parity with the existing Dataset.

### How was this patch tested?
Added a lot of golden tests.

Added a number of test cases to the E2E suite for the functionality that requires server interaction.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @zhenlineo

**Comment by hvanhovell** [comment]
For the reviewers. This is a mostly mechanical PR; the size is large but the complexity is low. All implemented documentation and function signatures were copies from Dataset.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -95,6 +97,121 @@ class Column private[sql] (private[sql] val expr: proto.Expression) extends Logg
...
+   * {{{
+   *   // Scala
+   *   df.sort(df("age").desc)
+   *
+   *   // Java
+   *   df.sort(df.col("age").desc());
+   * }}}
+   *
+   * @group expr_ops
+   * @since 1.3.0
```
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:112]: (should probably fix these version info)

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Column.scala:112]: yeah will do

**Review by HyukjinKwon — APPROVED** [review]

**Comment by hvanhovell** [comment]
merging this!

**Pull request closed without merging**